webrender/src/device/gfx/device.rs

```diff
@@ -1152,23 +1281,85 @@ impl<B: hal::Backend> Device<B> {
         }
     }
 
+    pub fn set_uniforms(&mut self, program: &ProgramId, projection: &Transform3D<f32>) {
+        debug_assert!(self.inside_frame);
+        assert_ne!(self.bound_program, INVALID_PROGRAM_ID);
+        assert_eq!(*program, self.bound_program);
+
+        let locals = Locals {
+            uTransform: projection.to_row_arrays(),
+            uMode: self.program_mode_id,
+        };
+
+        let bound_locals = if self.locals.contains_key(&locals) {
+            *self.locals.get(&locals).unwrap()
+        } else {
+            assert!(self.desc_pool_locals.next());
+            let (descriptor_set, idx) = self.desc_pool_locals.descriptor_set();
+            self.locals_buffer.add(&self.device, &[locals], &mut self.heaps);
```

Comment:
It's unsound for the API to separate `add()` from `buffer()`, given that pretty much nothing is allowed to go in-between. Can we just return a buffer reference from `add()` instead?